### PR TITLE
[shardformer] add Dropout layer support different dropout pattern

### PR DESCRIPTION
**Pull request by FoolPlayer**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

1. add dropout layer to support different drop pattern when training, and just modify the local random state when dropout 
2. add dropout test in test.py

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/layer/dropout.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import torch
+import torch.nn as nn
+
+
+class SeedManager:
+
+    def __init__(self):
+        self.original_state = torch.cuda.get_rng_state()
+        seed = int(f"{int(time.time())}{os.environ['RANK']}")
+        print(seed)
```
Comment (FrankLeeeee) [inline colossalai/shardformer/layer/dropout.py:13]: remove this line

File: colossalai/shardformer/layer/dropout.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        self.dropout_state = torch.cuda.get_rng_state()
+        torch.cuda.set_rng_state(self.original_state)
+
+    def dropout_mode(self):
+        self.original_state = torch.cuda.get_rng_state()
+        torch.cuda.set_rng_state(self.dropout_state)
+
+    def origin_mode(self):
+        self.dropout_state = torch.cuda.get_rng_state()
+        torch.cuda.set_rng_state(self.original_state)
```
Comment (FrankLeeeee) [inline colossalai/shardformer/layer/dropout.py:24]: wrap this with context manager.

File: colossalai/shardformer/layer/dropout.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+class Dropout1D(nn.Dropout):
+
+    def __init__(self, p=0.5, inplace=False):
+        super().__init__(p, inplace)
+
+    def forward(self, input):
+        _seed_manager.dropout_mode()
+        input = super().forward(input)
+        _seed_manager.origin_mode()
```
Comment (FrankLeeeee) [inline colossalai/shardformer/layer/dropout.py:38]: ideal way of writing this code

```python
with _seed_manager.dropout_mode():
        input = super().forward(input).

```

File: colossalai/shardformer/shard/slicer.py [hunk 4]
```diff
@@ -94,10 +94,11 @@ def slice_1d(
...
         """
-        delta = (tensor.shape[0] + self.shardconfig.world_size - 1) // self.shardconfig.world_size
-        down_idx = self.shardconfig.rank * delta
-        up_idx = down_idx + delta
-        return tensor[down_idx:up_idx].contiguous()
+        return tensor.chunk(self.shardconfig.world_size, dim=0)[self.shardconfig.rank].contiguous()
+        # delta = (tensor.shape[0] + self.shardconfig.world_size - 1) // self.shardconfig.world_size
+        # down_idx = self.shardconfig.rank * delta
+        # up_idx = down_idx + delta
+        # return tensor[down_idx:up_idx].contiguous()
```
Comment (FrankLeeeee) [inline colossalai/shardformer/shard/slicer.py:101]: Remove unused code.

File: colossalai/shardformer/shard/slicer.py [hunk 5]
```diff
@@ -113,10 +114,11 @@ def slice_col(
...
         """
-        delta = (tensor.shape[0] + self.shardconfig.world_size - 1) // self.shardconfig.world_size
-        down_idx = self.shardconfig.rank * delta
-        up_idx = down_idx + delta
-        return tensor[down_idx:up_idx, :].contiguous()
+        return tensor.chunk(self.shardconfig.world_size, dim=0)[self.shardconfig.rank].contiguous()
+        # delta = (tensor.shape[0] + self.shardconfig.world_size - 1) // self.shardconfig.world_size
+        # down_idx = self.shardconfig.rank * delta
+        # up_idx = down_idx + delta
+        # return tensor[down_idx:up_idx, :].contiguous()
```
Comment (FrankLeeeee) [inline colossalai/shardformer/shard/slicer.py:121]: same

**Review by FrankLeeeee — APPROVED** [review]

**Review by FrankLeeeee — APPROVED** [review]

**Review by FrankLeeeee — APPROVED** [review]
